mongosql/src/mir/optimizer/stage_movement/mod.rs

```diff
@@ -36,19 +36,24 @@ use crate::{
 use std::collections::HashSet;
 
 impl Stage {
-    // This is used for moving Offset as high as possible. They can be moved ahead of Any
-    // Stage that is not defined as offset invalidating.
-    fn is_offset_invalidating(&self) -> bool {
+    /// `is_limit_or_offset_invalidating` is used for moving Limit and Offset stages
+    /// as high as possible. They can be moved ahead of any Stage that
+    /// is not defined as invalidating.
+    fn is_limit_or_offset_invalidating(&self) -> bool {
         match self {
             // A tautological Filter will not invalidate offset, but we don't have a SAT solver.
             Stage::Filter(_) => true,
             Stage::Project(_) => false,
             // It's possible to have a Group that does not modify cardinality and thus invalidate
             // an offset, but we can consider that a very rare occurrence.
             Stage::Group(_) => true,
+            // ordering of two Limits does not matter. Limit 5, Limit 3 = Limit 3, Limit 5.
+            // MongoDB coalesces two Limits into one
             Stage::Limit(_) => true,
             // ordering of two Offsets does not matter. Offset 5, Offset 3 = Offset 3, Offset 5.
-            Stage::Offset(_) => false,
+            // However, Limit should not be moved above an Offset.
+            // MongoDB coalesces two Offsets($skip) into one
```

Comment:
These comments are incomplete.

Not only do we not want to move a `Limit` above an `Offset`, but also we don't want to move an `Offset` above a `Limit`. For example, for `SELECT * FROM (SELECT a FROM foo LIMIT 10) OFFSET 5` we could push the `OFFSET 5` into the derived query but it actually still come after the `LIMIT 10` or else we'll change the result set.

Can we update these comments to point out both cases?

Reply:
I've updated the function comment since it seemed appropriate to do so there!